### PR TITLE
Don't retry unretryable errors

### PR DESCRIPTION
**Pull request by gdbelvin**
The current `Retry()` logic loops forever on all errors. 
In practice, this is creating infinite loops when things don't go well. 

This PR makes `Retry()` a bit more intelligent by only retrying retryable and unknown errors.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Adding @pavelkalinnikov as this will affect Migrillian / scanner use of this code.

File: client/backoff/backoff.go [hunk 1]
```diff
@@ -80,3 +83,45 @@ func (b *Backoff) Retry(ctx context.Context, f func() error) error {
 		return nil
 	}
 }
+
+// IsRetryable returns false if the error code indicates
```
Comment (pavelkalinnikov) [inline client/backoff/backoff.go]: Suggestion: shorten the whole comment to:
```
// IsRetryable returns true unless the error is explicitly not retriable per
// https://godoc.org/google.golang.org/grpc/codes.
```

File: client/backoff/backoff.go [hunk 2]
```diff
@@ -67,9 +70,9 @@ func (b *Backoff) Retry(ctx context.Context, f func() error) error {
 		return ctx.Err()
 	}
 
-	// Try calling f until it doesn't return an error or ctx is done.
+	// Try calling f while the error is retryable and ctx is not done.
 	for {
-		if err := f(); err != nil {
+		if err := f(); IsRetryable(err) {
```
Comment (pavelkalinnikov) [inline client/backoff/backoff.go]: A reasonable next step would be allowing the client to specify a custom "is retriable" predicate, e.g. for non-gRPC.
Maybe TODO?

Comment (gdbelvin) [reply client/backoff/backoff.go]: My personal thought is that if clients should implement the retry loop themselves and call `b.Duration()` directly in order to handle any complex retry logic.

Comment (pavelkalinnikov) [reply client/backoff/backoff.go]: SGTM. Also, they can do smth like [this](https://github.com/google/certificate-transparency-go/blob/04d4fbb06517a10291ff56ddbc11aa7390b8393a/trillian/migrillian/core/trillian.go#L123).

**Pull request merged**